Analysis/DataStorageProjects/Delfin/CentralizedManager/CentralizedManager.md

```diff
@@ -0,0 +1,140 @@
+
+  Delete shall report failed status for invalid CM ids
+
+- Users of delfin should be able to list all registered CMs details
+
+  List shall report all the registered CMs and its properties excluding sensitive details like password
+
+  If no CM is registered, the listing will return an empty list
+
+- Users of delfin should be able to show CM details for a registered CM, providing its id
+
+  Show shall get CM details for the id of the CM, that is currently managed by delfin
+
+  Invalid id for CM shall report error
+
+- Users of delfin should be able to update CM Access Info details for a registered CM, providing its id
+
+  Update shall verify the Access Info and if verification is success, update the CM details for the id of the CM, that is currently managed by delfin
+
+  Invalid Access Info for CM shall report error and no changes shall be done to delfin
+
+- Users of delfin should be able to SYNC all registered CMs for updating list of managed storage arrays of each CM
+
+  When user calls SYNC all, delfin with re-discover all the storage arrays from CM and update managed storage arrays
+
+- Users of delfin should be able to SYNC a specific CM with the its id, for updating list of managed storage arrays by that CM
+
+  When user calls SYNC for a CM, delfin with re-discover all the storage arrays from the CM and update managed storage arrays
+
+### Non Functional Requirements
+
+- Centralized Manager feature needs to be generic to support different storage vendors
+- Centralized Manager feature needs to be compatible with already existing APIs of Delfin
+
+## Architecture Analysis
+
+Adding support for CM requires the addition of CM related APIs to current delfin APIs. Driver Manager and Driver API will require update. Impacted modules of delfin are highlighted below
+
+![Centralized Managers Architecture](Delfin.png)
+
+### High Level Design
+
+Add the Centralized Manager as a separate entity with APIs to support Create, Delete and Get operations. Access Info for CM can be with different Schema and supports APIs Update/Get. APIs for CM sync and sync-all also will be added for updating managed storages by CMs.
+
+- Add REST APIs to support CRUD operations of CM and its access info.
+- Update framework, driver manager APIs for supporting CM drivers
+- None of the existing REST APIs will be impacted, existing delfin users will not require changes for existing features
+
+#### Access Info model changes for CM
+
+```sh
+{
+  <same as existing access_info>
+  “array_ids”: [<storage_uuid>, ...]
+  “centralized_manager_id”: <manager_uuid>
+}
+```
+
+#### New APIs for CM
+
+```sh
+POST /centralized-managers -d <cm-access-info>
+
+GET /centralized-managers
+
+GET /centralized-managers/<cm-id>
+
+PUT /centralized-managers/access-info -d <cm-access-info>
+
+GET /centralized-managers/access-info
+
+PUT /centralized-managers/sync-all
+
+PUT /centralized-managers/<cm-id>/sync
+```
+
+#### Common Access Info for storage managers
```

Comment:
Is there any exisiting user impact ? I hope since the DB model has a change user has to recreate the DB or migrate to new DB .. we should mention that.